### DISCUSSION TITLE
Rendering post titles on the server side.

### DISCUSSION THREAD

**Discussion by SabrioBiblio** [vercel/next.js#62509] [Help]
### Summary

I have a tricky (possibly incorrect) post system in next js 14.

```
post/
└── some-post/
    ├── page.tsx
    └── meta.md
```

The **meta.md** file stores the description, title and image of the post.
I wrote a script called **posts.ts** that reads this file and extracts information from it.

But there is one problem. On the server side I am unable to read this file correctly and send the correct post to the user.

There is a **layout.tsx** file in the post folder. It has code

```tsx
export default async function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  const fullUrl = headers().get("referer");
  const postSlug = fullUrl?.split("/").slice(-1)[0] || "";

  const post = await getPost(postSlug);

  return (
    <main>
      <PostHeader title={post?.title} date={post?.date} />
      {children}
    </main>
  );
}
```
But it feels like the server doesn’t wait for posts.ts (**_getPost()_** function) to be executed and skips it, and I see that the server returned empty space in the PostHeader component.

I'm sure the problem is me and I just don't understand how this should work correctly on the server side
​

**Comment by icyJoseph** [comment]
Hi,

Have you verified that `fullUrl` is correct? that `postSlug` is correct? Could you share the `getPost` implementation?

**Reply by SabrioBiblio** [reply]
Hi!
I am confident (**read comment below**) that I am getting the correct fullUrl. I also get the correct slugUrl.

**posts.ts -> getPost()**
```tsx
export const getPost = async (slug: string): Promise<Post | null> => {
  const postDir = fs.readdirSync(
    path.resolve(postsDirectoryPath, pathToPosts + "/" + slug)
  );
  const metaPostFile = postDir.filter((file) => file === metaFileName)[0];

  if (!metaPostFile) {
    return null;
  }

  const fullPath = path.join(
    postsDirectoryPath,
    pathToPosts,
    slug,
    metaPostFile
  );

  const postMeta = fs.readFileSync(fullPath, "utf8");
  const matterResult: any = await matter(postMeta);

  return {
    slug: slug,
    ...matterResult.data,
  };
};
```

**Reply by SabrioBiblio** [reply]
I just noticed one mistake. 

When I go to the post page, the **fullUrl** returns me simply **localhost:3000/**. But after hydration it returns me the full **localhost:3000/post/some-post**. I tried using regular routing and dynamic routing, the result is the same. This is a bit strange for me and I don't quite understand why it is so difficult

yes, headers().get("referer") definitely returns me localhost:3000. It's wrong

**Reply by SabrioBiblio** [reply]
Then the problem becomes more obvious, now all I need to do is get the current browser path, but as I understand it, it’s not that simple

**Reply by icyJoseph** [reply]
I'm out of office now, but could you try to run an experiment?

See, https://github.com/vercel/next.js/discussions/61818#discussioncomment-8407779

Create a Middleware like that, but instead of a cookie, put the req.nextUrl.pathname into a header, call it whatever you like, and try to read it from the layout.

**Reply by SabrioBiblio** [reply]
Yes, I already tried to do this, but in the layout component I get null. 
[#46618](https://github.com/vercel/next.js/issues/46618#issuecomment-1450416633)

```tsx
//midleware.ts
export function middleware(request: NextRequest) {
  return NextResponse.next({
     request: {
        // New request headers
        'x-pathname': request.nextUrl.pathname,
     },
  });
}

//layout.tsx
export default async function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  const fullUrl = headers().get("x-pathname"); // return null

  return (
    <main>
      {children}
    </main>
  );
}
```

**Reply by SabrioBiblio** [reply]
I've corrected the code a little changing request to header. Now this started working but client routing is not working correctly, it is showing previous post data. It seems to me that the x-pathname value is cached somewhere. I need client routing


```tsx
export function middleware(request: NextRequest) {
  return NextResponse.next({
    headers: {
      "x-pathname": request.nextUrl.pathname,
    },
  });
}
```

**Reply by icyJoseph** [reply]
Right, you had it wrong first. And then the issue is that you almost need a full client side RSC cache bust on every route transition 🤔 like, for some RSC that's not path dependent the serialized VDOM is kept client side and bust every now and then IIRC.

Not sure there's a good way here, try to rwad the Cache and Revalidation Deep Dive on this forum. I'll also think about it when I'm back at work.

**Reply by SabrioBiblio** [reply]
Still, the question is relevant.

setting an entry in headers is not the right solution, since it eliminates the possibility of using it in client routing.

I have static nested pages
post/
└──about-company/
 │└── page.tsx
└──about-dogs/
 │└── page.tsx
└──layout.tsx

I still don’t understand how I can get the current slug (post title) of the page in the layout.tsx component without using dynamic routing
